File: Android/Physio/app/src/main/java/com/ibm/mil/readyapps/physio/models/Patient.java
/*
 * Licensed Materials - Property of IBM
 * © Copyright IBM Corporation 2014, 2015. All Rights Reserved.
 */

package com.ibm.mil.readyapps.physio.models;

import org.json.JSONException;
import org.json.JSONObject;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class Patient {

    private static final String USER_ID = "userID";
    private static final String DATE_OF_NEXT_VISIT = "dateOfNextVisit";
    private static final String VISITS_USED = "visitsUsed";
    private static final String VISITS_TOTAL = "visitsTotal";
    private static final String STEPS_GOAL = "stepGoal";
    private static final String CALORIES_GOAL = "calorieGoal";

    private String userID;
    private Date dateOfNextVisit;
    private Integer visitsUsed;
    private Integer visitsTotal;
    private Integer stepGoal;
    private Integer calorieGoal;

    private List<Form> formList;
    private List<Routine> routineList;
    private Map<Routine, List<Exercise>> exerciseMap;
    private int exerciseSets = 0;

    private Routine selectedRoutine;
    private Exercise selectedExercise;

    /**
     * Initializes a Patient object with the data returned from Worklight
     *
     * @param jsonResp The JSONObject from Worklight
     */
    public Patient(JSONObject jsonResp) {
        try {
            String stringResult = jsonResp.getString("result");
            JSONObject jsonResult = new JSONObject(stringResult);

            this.setUserID(jsonResult.getString(USER_ID));
            this.visitsUsed = jsonResult.getInt(VISITS_USED);
            this.visitsTotal = jsonResult.getInt(VISITS_TOTAL);
            this.stepGoal = jsonResult.getInt(STEPS_GOAL);
            this.calorieGoal = jsonResult.getInt(CALORIES_GOAL);
            String dateOfNextVisitString = jsonResult.getString(DATE_OF_NEXT_VISIT);
            this.dateOfNextVisit = new SimpleDateFormat("MM/dd/yyyy", Locale.ENGLISH)
                    .parse(dateOfNextVisitString);
            formList = new ArrayList<>();
            routineList = new ArrayList<>();
            exerciseMap = new HashMap<>();
        } catch (JSONException|ParseException e) {
            e.printStackTrace();
        }
    }

    /**
     * Gets the userID of the Patient.
     *
     * @return String userID
     */
    public String getUserID() {
        return this.userID;
    }

    /**
     * Sets the userID of a Patient.
     *
     * @param userID
     */
    public void setUserID(String userID) {
        this.userID = (userID != null) ? userID : "";
    }


    /**
     * Gets the date of the next visit for the Patient.
     *
     * @return Date dateOfNextVisit
     */
    public Date getDateOfNextVisit() {
        return this.dateOfNextVisit;
    }


    /**
     * Gets the number of visits the Patient has used.
     *
     * @return Integer visitsUsed
     */
    public Integer getVisitsUsed() {
        return this.visitsUsed;
    }


    /**
     * Gets the total number of visits the Patient has.
     *
     * @return Integer visitsTotal
     */
    public Integer getVisitsTotal() {
        return this.visitsTotal;
    }

    /**
     * The step goal for the patient.
     *
     * @return Integer stepGoal
     */
    public Integer getStepGoal() {
        return this.stepGoal;
    }


    /**
     * The calorie goal for the patient.
     *
     * @return Integer calorieGoal
     */
    public Integer getCalorieGoal() {
        return this.calorieGoal;
    }

    /**
     *
     * @param formList A list of form questions associated with the patient
     */
    public void setForms(List<Form> formList) {
        this.formList = formList;
    }

    /**
     *
     * @return A list of form questions associated with the patient
     */
    public List<Form> getForms() {
        return formList;
    }

    /**
     *
     * @param routineList A list of routines associated with the patient
     */
    public void setRoutines(List<Routine> routineList) {
        this.routineList = routineList;
    }

    /**
     *
     * @return A list of routines associated with the patient
     */
    public List<Routine> getRoutines() {
        return routineList;
    }

    /**
     *
     * @param routine The routine that the exercises fall under
     * @param exerciseList The list of exercises for the specified routine
     */
    public void addExercises(Routine routine, List<Exercise> exerciseList) {
        exerciseMap.put(routine, exerciseList);
        exerciseSets++;
    }

    /**
     *
     * @param routine The routine that the desired exercises belong to
     * @return The list of exercises for the specified routine
     */
    public List<Exercise> getExercises(Routine routine) {
        return exerciseMap.get(exerciseMap.keySet().iterator().next());
    }

    /**
     *
     * @param routine The routine selected by the user in the RoutineFragment
     */
    public void setSelectedRoutine(Routine routine) {
        selectedRoutine = routine;
    }

    /**
     *
     * @return The routine selected by the user in the RoutineFragment
     */
    public Routine getSelectedRoutine() {
        return selectedRoutine;
    }

    /**
     *
     * @param exercise The exercise selected by the user in the ExerciseFragment
     */
    public void setSelectedExercise(Exercise exercise) {
        selectedExercise = exercise;
    }

    /**
     *
     * @return The exercise selected by the user in the ExerciseFragment
     */
    public Exercise getSelectedExercise() {
        return selectedExercise;
    }

    /**
     *
     * @return The number of Routines that contain exercises
     */
    public int getNumberOfExerciseSets() {
        return exerciseSets;
    }

}

File: iOS/ReadyAppPT/Models/Patient.swift
/*
Licensed Materials - Property of IBM
© Copyright IBM Corporation 2014, 2015. All Rights Reserved.
*/

import UIKit

/**
Patient is a model of data representing the current user of the app. This is an extension of IBMDataObject
*/
class Patient: NSObject {
   
     var userID : String!
     var password : String!
     var dateOfNextVisit : NSDate!
     var visitsUsed : NSNumber!
     var visitsTotal : NSNumber!
     var stepGoal : NSNumber!
     var calorieGoal : NSNumber!
     var routineId : [String]!
     var questionnaire : String!
   
    /**
    An alternative init method that allows for populating all the properties of the object
    
    :param: userID          The ID of the patient
    :param: password        The password for the patient
    :param: dateOfNextVisit NSDate object representing the date of the next visit to doctor
    :param: visitsUsed      The number of visits the patient has used on visit plan
    :param: visitsTotal     The number of visits the patient has planned with doctor in total
    :param: stepGoal     The step goal for the patient as a number
    :param: calorieGoal     The calorie goal for the patient as a number
    
    */
    init(userID : String, password : String, dateOfNextVisit: NSDate, visitsUsed: NSNumber, visitsTotal: NSNumber, stepGoal : NSNumber, calorieGoal : NSNumber, routineId : [String]!, questionnaire : String!) {
        super.init()
        self.userID = userID
        self.password = password
        self.dateOfNextVisit = dateOfNextVisit
        self.visitsUsed = visitsUsed
        self.visitsTotal = visitsTotal
        self.stepGoal = stepGoal
        self.calorieGoal = calorieGoal
        self.routineId = routineId
        self.questionnaire = questionnaire
    }
    
    
    init(worklightResponseJson: NSDictionary) {
        super.init()
        
        
        let jsonResult = worklightResponseJson["result"] as! NSDictionary
        /*
        let stringResult = worklightResponseJson["result"] as! String
        let data = stringResult.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)
        let jsonResult = NSJSONSerialization.JSONObjectWithData(data!, options: NSJSONReadingOptions.MutableLeaves, error: nil) as! NSDictionary
        */
        
        self.userID = jsonResult["userID"] as! String
        self.visitsUsed = jsonResult["visitsUsed"] as! NSNumber
        self.visitsTotal = jsonResult["visitsTotal"] as! NSNumber
        self.stepGoal = jsonResult["stepGoal"] as! NSNumber
        self.calorieGoal = jsonResult["calorieGoal"] as! NSNumber
        
        let dateOfNextVisitString = jsonResult["dateOfNextVisit"] as! String
        let formatter = NSDateFormatter()
        formatter.dateFormat = "MM/dd/yyyy"
        self.dateOfNextVisit = formatter.dateFromString(dateOfNextVisitString)
        self.routineId = jsonResult["routines"] as! [String]
        self.questionnaire = jsonResult["questionnaire"] as! String
    }
    

}
